transport_plugins/bled112/iotile_transport_bled112/hardware/async_bled112.py

```diff
@@ -0,0 +1,459 @@
+        """Probe for all primary services and characteristics in those services.
+
+        Args:
+            handle (int): the connection handle to probe
+        """
+
+        end_spec = MessageSpec(class_=4, cmd=1, conn=handle)
+        gatt_spec = MessageSpec(class_=4, cmd=2, conn=handle)
+        payload = struct.pack('<BHHBH', handle, 1, 0xFFFF, 2, 0x2800)
+
+        async with self._locks.conn(handle):
+            await self.send_command_locked(4, 1, payload, pause=True, check="<xH")
+            events = await self.operations.gather_until([gatt_spec, end_spec], end_spec, timeout=2.0, unpause=True)
+
+        end_event = events[-1]
+        gatt_events = events[:-1]
+
+        #Make sure we successfully probed the gatt table
+        _, result, _ = struct.unpack("<BHH", end_event.payload)
+        if result != 0:
+            self._logger.warning("Error enumerating GATT table, protocol error code = %d (0x%X)", result, result)
+            return False, None
+
+        services = {}
+        for event in gatt_events:
+            process_gatt_service(services, event)
+
+        return {'services': services}
+
+    async def probe_characteristics(self, conn, services, timeout=5.0):
+        """Probe gatt services for all associated characteristics in a BLE device.
+
+        Args:
+            conn (int): the connection handle to probe
+            services (dict): a dictionary of services produced by probe_services()
+            timeout (float): the maximum number of seconds to spend in any single task
+        """
+
+        for service in services.values():
+            result = await self._enumerate_handles(conn, service['start_handle'], service['end_handle'])
+
+            attributes = result['attributes']
+
+            service['characteristics'] = {}
+
+            last_char = None
+            for handle, attribute in attributes.items():
+                if attribute['uuid'].hex[-4:] == '0328':
+                    result = await self._read_handle(conn, handle, timeout)
+
+                    value = result['data']
+                    char = parse_characteristic_declaration(value)
+                    service['characteristics'][char['uuid']] = char
+                    last_char = char
+                elif attribute['uuid'].hex[-4:] == '0229':
+                    if last_char is None:
+                        raise DeviceAdapterError(None, 'probe_characteristics',
+                                                 'invalid gatt table that was not ordered correctly')
+
+                    result = await self._read_handle(conn, handle, timeout)
+
+                    value = result['data']
+                    assert len(value) == 2
+                    value, = struct.unpack("<H", value)
+
+                    last_char['client_configuration'] = {'handle': handle, 'value': value}
+
+        return {'services': services}
+
+    async def write_highspeed(self, conn, handle, data, progress_callback=None):
```

Comment:
I believe `handle` is supposed to be an `int`, based on how `_write_handle` is using it. You mentioned that this is mostly a port of an old class, so it probably didn't have a docstring to begin with.

Reply:
It is.  I'm going to add the type annotation later here because I think this method will actually get removed from this class before the final merge of this branch as I see how the next phase is shaping up to look.